### PR TITLE
Data export for v2 REST-API

### PR DESCRIPTION
**Pull request by ewelinagr**
Data export was so far a part of transmartApp. Now it is also implemented in a new way, as a part of transmart-rest-api. Basic functionality is the same - data export is run an asynchronous job, exported data is stored in a temporal directory as .zip file(directory can be change in configuration file). 

The file is created using HypercubeCSVSerializer, by parsing hypercube to TSV files and saving them to a zipOutputStream. 
New REST-API endpoints created:
* POST /v2/export/job?name={name}
** name - optional
* POST /v2/export/{jobId}/run?typeOfSet={typeOfSet}&id={id1}&id={id2}&id=...&elements={elements}
** {typeOfSet} can be one of ['observation', 'patient']
** {id} is a list of patient_sets/observations_sets (observation_sets not supported for now)
** {elements} - list of pairs: {dataType:{dataType}, format:(fileFormat}, JSON format
* GET /v2/export/{jobId}/download
* GET /v2/export/{jobId}/status
* GET /v2/export/jobs
* GET /v2/export/data_formats?id={id1}&id={id2}&id=...&typeOfSet={typeOfSet}
* GET /v2/export/file_formats

### CODE REVIEW

**Comment by ewelinagr** [comment]
Integration tests still need to be added.

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            newJob.save()
+        }
+
+        newJob
+    }
+    
+    /**
+     * Cancel a running job
+     */
+    def cancelJob(jobName, group = null) {
```
Comment (ewelinagr) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy]: This is not used for now, but we may want to have function like this at some point.

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    Scheduler quartzScheduler
+
+    static String jobType = 'DataExport'
+
+    /**
+     * Get the list of jobs to show in the jobs tab
+     */
+    List<AsyncJob> getJobList(User user) {
+        // TODO add time constraints for retrived jobs
```
Comment (ewelinagr) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportAsyncJobService.groovy]: For now we agreed that all the jobs will be received, later we should add some time limit (configurable) parameter

File: transmart-rest-api/src/main/groovy/org/transmartproject/rest/dataExport/ExportJobExecutor.groovy [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+//    QuartzSpringScope quartzSpringScope = ctx.quartzSpringScope
+
+    final String tempFolderDirectory = Holders.config.com.recomdata.plugins.tempFolderDirectory
+
+    public void execute(JobExecutionContext jobExecutionContext) {
+        def userInContext = jobExecutionContext.jobDetail.jobDataMap['userInContext']
+        Map jobDataMap = jobExecutionContext.jobDetail.getJobDataMap()
+
+//        // put the user in context
+//        quartzSpringScope.currentUserBeanQuartzScope = userInContext
```
Comment (ewelinagr) [inline transmart-rest-api/src/main/groovy/org/transmartproject/rest/dataExport/ExportJobExecutor.groovy]: This was a part of transmartApp async job executor, I am not sure why and if it is needed (also setting currentUserBeanQuartzScope). @JanKanis, @forus do you know, if something with similar to [QuartzSpringScope](https://github.com/thehyve/transmart-core/blob/master/transmartApp/src/main/groovy/org/transmart/spring/QuartzSpringScope.groovy) from tm-app should be added here?

File: transmart-rest-api/grails-app/controllers/org/transmartproject/rest/RestExportController.groovy [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+     * @return {@link AsyncJob} instance
+     */
+    def createJob(@RequestParam('jobName') String jobName) {
+        checkParams(params, ['jobName'])
+
+        User user = (User) usersResource.getUserFromUsername(currentUser.username)
+        checkJobNameUnique(jobName)
+
+        def instance = restExportService.createExportJob(user, jobName)
+        respond instance
```
Comment (ewelinagr) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/rest/RestExportController.groovy]: Maybe it is better to add some wrapper here, but I am not sure what exactly need to be returned.

File: transmart-rest-api/src/main/groovy/org/transmartproject/rest/dataExport/ExportJobExecutor.groovy (hunk 3 above)
Comment (JanKanis) [inline transmart-rest-api/src/main/groovy/org/transmartproject/rest/dataExport/ExportJobExecutor.groovy]: If the user is to be injected in the background job something similar will be needed. But I don't know if the user needs to be injected and i would say preferrably not (though I don't know what the spring best practices would be in such a context).

File: transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy [hunk 5]
```diff
@@ -827,13 +818,46 @@ class MultidimensionalDataResourceService implements MultiDimensionalDataResourc
         }
     }
 
+    @Override List retriveHighDimDataTypes(MultiDimConstraint assayConstraint_, User user){
```
Comment (JanKanis) [inline transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy]: Don't we already have this functionality somewhere?

**Comment by forus** [comment]
> POST /v2/export/job?jobName={jobName}
> ** jobName - optional
if the job name is used as id I would vote for generating it on the fly (and call it id), don't give possibility to a user to specify it.

> GET /v2/export/{jobName}/run?setType={setType}&ids={ids}&elements={elements}
> ** {set_type} can be one of ['observation', 'patient']
> ** {ids} is a list of patient_sets/observations_sets (observation_sets not supported for now)
> ** {elements} - list of pairs: {dataType:{dataType}, format:(fileFormat}, JSON format
it would look more consistent to have `../job/{jobName}/run`.
This call should use `POST` method. We might want to pass parameters in the request body.

> GET /v2/export/{jobName}/download
it would look more consistent to have `../job/{jobName}/download`

> GET /v2/export/{jobName}/status
it would look more consistent to have `../job/{jobName}/status`

> GET /v2/export/data_formats/{ids}?setType={setType}
How different the data fromat from the file format?
> GET /v2/export/supported_file_formats
Does prefix `supported_` add up any semantic value? Wouldn't `../file_format` be good enough?

File: transmart-rest-api/src/test/groovy/org/transmartproject/rest/RestExportServiceSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import spock.lang.Specification
+
+/**
+ * See the API for {@link grails.test.mixin.services.ServiceUnitTestMixin} for usage instructions
+ */
+@TestFor(RestExportService)
+class RestExportServiceSpec extends Specification {
+//
+//    @Autowired
+//    RestDataExportService restDataExportService
```
Comment (forus) [inline transmart-rest-api/src/test/groovy/org/transmartproject/rest/RestExportServiceSpec.groovy]: Please remove commented code.

File: transmart-rest-api/src/test/groovy/org/transmartproject/rest/RestExportServiceSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    
+    def setup() {
+    }
+
+    def cleanup() {
+    }
+
+    void "test something"() {
```
Comment (forus) [inline transmart-rest-api/src/test/groovy/org/transmartproject/rest/RestExportServiceSpec.groovy]: ?

File: transmart-rest-api/src/main/groovy/org/transmartproject/rest/dataExport/JobStatus.groovy [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package org.transmartproject.rest.dataExport
+
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+
+@CompileStatic
+@Slf4j
```
Comment (forus) [inline transmart-rest-api/src/main/groovy/org/transmartproject/rest/dataExport/JobStatus.groovy]: No log statements in the enum. You could remove `@Slf4j`.

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestExportService.groovy [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+        def jobDataMap = new JobDataMap(dataMap)
+        JobDetail jobDetail = JobBuilder.newJob(ExportJobExecutor.class)
+                .withIdentity(dataMap.jobName, 'DataExport')
+                .setJobData(jobDataMap)
+                .build()
+
+        exportAsyncJobService.updateStatus(dataMap.jobName, JobStatus.TRIGGERING_JOB)
+
+        def randomDelay = Math.random()*10 as int
```
Comment (forus) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestExportService.groovy]: Why the delay time has random component?

File: transmart-rest-api/build.gradle [hunk 10]
```diff
@@ -35,8 +35,11 @@ dependencies {
     testCompile "org.grails:grails-plugin-testing"
     testRuntime "net.sourceforge.htmlunit:htmlunit:2.18"
     testCompile 'org.grails:grails-datastore-rest-client'
+    compile 'org.quartz-scheduler:quartz:2.2.1'
+    compile 'net.sf.opencsv:opencsv:2.3'
 
     compile 'org.grails.plugins:spring-security-core:3.1.1'
+    compile project(':transmart-legacy-db')
```
Comment (forus) [inline transmart-rest-api/build.gradle]: Why do we introduce dependency on `:transmart-legacy-db` project?

File: transmart-rest-api/grails-app/controllers/org/transmartproject/rest/RestApiUrlMappings.groovy [hunk 11]
```diff
@@ -133,6 +133,27 @@ class RestApiUrlMappings {
             "/recommended_concepts/$conceptCode"(method: 'GET', controller: 'concept', action: 'showRecommended') {
                 apiVersion = 'v2'
             }
+            "/export/job"(method: 'POST', controller: 'restExport', action: 'createJob') {
```
Comment (forus) [inline transmart-rest-api/grails-app/controllers/org/transmartproject/rest/RestApiUrlMappings.groovy]: It is preferable to follow Grails conventions, don't have custom url mappings.

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestDataExportService.groovy [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+import org.transmartproject.db.multidimquery.query.PatientSetConstraint
+import org.transmartproject.rest.MultidimensionalDataService
+import org.transmartproject.rest.MultidimensionalDataService.Format as Format
+import sun.reflect.generics.reflectiveObjects.NotImplementedException
+
+import java.util.zip.ZipOutputStream
+
+
+@Transactional
+class RestDataExportService {
```
Comment (forus) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestDataExportService.groovy]: What makes this service `Rest` ?  Shouldn't it be simply `DataExportService`?

File: transmart-rest-api-e2e/src/test/groovy/tests/rest/v2/DataExportSpec.groovy [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+        String fileName = "$TEMP_DIRECTORY/$jobName"+".zip"
+        def downloadRequest = [
+                path      : "$PATH_DATA_EXPORT/$jobName/download",
+                acceptType: contentTypeForZip,
+        ]
+        def downloadResponse = get(downloadRequest)
+
+        then: "ZipStream is returned"
+        assert downloadResponse != null
+        assert new File(fileName).isFile()
```
Comment (forus) [inline transmart-rest-api-e2e/src/test/groovy/tests/rest/v2/DataExportSpec.groovy]: It would be better to test response content, not a file on the file system.

File: transmart-rest-api/build.gradle (hunk 10 above)
Comment (ewelinagr) [inline transmart-rest-api/build.gradle]: AsyncJob domain class is a part of transmart-legacy-db

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestExportService.groovy (hunk 9 above)
Comment (ewelinagr) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestExportService.groovy]: This part is copied from transmartApp/ExportService.groovy and was added by Gijs to fix some issue with scheduler after Grails 3 upgrade

File: transmart-rest-api/src/test/groovy/org/transmartproject/rest/RestExportServiceSpec.groovy (hunk 7 above)
Comment (ewelinagr) [inline transmart-rest-api/src/test/groovy/org/transmartproject/rest/RestExportServiceSpec.groovy]: This test was auto-generated by Grails. I didn't remove it, because I am not sure if we need tests for RestExportService, they will partially duplicate e2e tests.

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestDataExportService.groovy (hunk 12 above)
Comment (ewelinagr) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestDataExportService.groovy]: DataExportService exists already for transmartApp and Spring is not always handling this kind of duplicates correctly. I thought this is safer solution, but I can try to resolve this.

File: transmart-rest-api-e2e/src/test/groovy/tests/rest/v2/DataExportSpec.groovy (hunk 13 above)
Comment (ewelinagr) [inline transmart-rest-api-e2e/src/test/groovy/tests/rest/v2/DataExportSpec.groovy]: The response content is already tested in [ObservationsBuilderTests](https://github.com/thehyve/transmart-core/blob/data_export/transmart-rest-api/src/test/groovy/org/transmartproject/rest/protobug/ObservationsBuilderTests.groovy#L182), as an integration test for HypercubeCSVSerializer. Should I check it here again?

File: transmart-rest-api/build.gradle (hunk 10 above)
Comment (forus) [inline transmart-rest-api/build.gradle]: We would better to duplicate them in the core-db and remove this dependency. See `*CoreDb.groovy` domain classes in core-db.

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestDataExportService.groovy (hunk 12 above)
Comment (forus) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/RestDataExportService.groovy]: it would be nice to resolve it. transmartApp will be out of the picture at some point.

File: transmart-rest-api/src/test/groovy/org/transmartproject/rest/RestExportServiceSpec.groovy (hunk 7 above)
Comment (forus) [inline transmart-rest-api/src/test/groovy/org/transmartproject/rest/RestExportServiceSpec.groovy]: I see. Please remove this class then.

File: transmart-rest-api-e2e/src/test/groovy/tests/rest/v2/DataExportSpec.groovy (hunk 13 above)
Comment (forus) [inline transmart-rest-api-e2e/src/test/groovy/tests/rest/v2/DataExportSpec.groovy]: IMHO no need to duplicate efforts in testing the content. It would be enough just to prove what is stated `ZipStream is returned`.

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportService.groovy [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+import org.springframework.stereotype.Component
+import org.transmartproject.core.exceptions.InvalidRequestException
+import org.transmartproject.db.job.AsyncJobCoreDb as AsyncJob
+import org.transmartproject.db.user.User
+
+import javax.transaction.NotSupportedException
+
+@Transactional
+@Component("restExportService")
+class ExportService {
```
Comment (forus) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportService.groovy]: There is a lot of logic shared with `DataExportService` and `ExportAsyncJobService`. Let's have a look whether we could reduce the number of services.

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportService.groovy [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+import org.transmartproject.core.exceptions.InvalidRequestException
+import org.transmartproject.db.job.AsyncJobCoreDb as AsyncJob
+import org.transmartproject.db.user.User
+
+import javax.transaction.NotSupportedException
+
+@Transactional
+@Component("restExportService")
```
Comment (ewelinagr) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportService.groovy]: yes, without this I was getting BeanCreationException: "Error creating bean with name 'org.transmartproject.rest.ExportController': Initialization of bean failed; nested exception is org.springframework.beans.ConversionNotSupportedException: Failed to convert property value of type [com.recomdata.transmart.data.export.ExportService] to required type [org.transmartproject.rest.dataExport.ExportService] for property 'exportService'; nested exception is java.lang.IllegalStateException: Cannot convert value of type [com.recomdata.transmart.data.export.ExportService] to required type [org.transmartproject.rest.dataExport.ExportService] for property 'exportService': no matching editors or conversion strategy found"

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportService.groovy [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+class ExportService {
+
+    @Autowired(required = false)
```
Comment (ewelinagr) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportService.groovy]: There is an issue with autowiring for some classes - e.g. ObservationsBuilderTests in rest-api were failing with exception:  org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'restExportService': Unsatisfied dependency expressed through field 'quartzScheduler'; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean found for dependency [org.quartz.Scheduler]: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Autowired(required=true)}

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportService.groovy [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+    private AsyncJob executeExportJob(Map dataMap) {
+
+        def job
+        def jobDataMap = new JobDataMap(dataMap)
+        JobDetail jobDetail = JobBuilder.newJob(ExportJobExecutor.class)
+                .withIdentity(dataMap.jobId.toString(), 'DataExport')
+                .setJobData(jobDataMap)
+                .build()
+        def randomDelay = Math.random()*10 as int
```
Comment (ewelinagr) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/ExportService.groovy]: I am not sure, this was added by Gijs in [f84542c](https://github.com/thehyve/transmart-core/commit/f84542cc19aa3db6ef02dbe026481ed2368ccfa1) commit to fix job schedulers in transmartApp

File: transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/DataExportService.groovy [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        throw new NotImplementedException()
+    }
+
+    private List highDimDataTypes(Constraint constraint, User user) {
+        multidimensionalDataService.multiDimService.retriveHighDimDataTypes(constraint, user)
+    }
+
+    List patientSetsExportPermission(List<Long> ids, User user){
+        ids.collect { setId ->
+            multidimensionalDataService.multiDimService.findPatientSet(setId, user, ProtectedOperation.WellKnownOperations.EXPORT)
```
Comment (ewelinagr) [inline transmart-rest-api/grails-app/services/org/transmartproject/rest/dataExport/DataExportService.groovy]: So why EXPORT is one of ProtectedOperation.WellKnownOperations? Isn't it different access level?

File: transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy [hunk 19]
```diff
@@ -831,13 +822,46 @@ class MultidimensionalDataResourceService implements MultiDimensionalDataResourc
...
+
+        Constraint assayConstraint = (Constraint) assayConstraint_
```
Comment (ewelinagr) [inline transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy]: I followed the convention which was used by Jan in other methods in this class. Should I refactor them as well?
